docs/configuration-reference.md
# Swoole Bundle Configuration

Documentation of available configuration parameters. See also symfony [bundle configuration](./../src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php) file or [swoole documentation](https://github.com/swoole/swoole-docs/tree/master/modules).

- [Swoole Bundle Configuration](#swoole-bundle-configuration)
  - [HTTP Server](#http-server)

## HTTP Server

*Example*:

```yaml
swoole:
    http_server:
        port: 9501
        host: 0.0.0.0
        running_mode: process
        socket_type: tcp
        ssl_enabled: false
        trusted_hosts: localhost,127.0.0.1
        trusted_proxies:
            - '*'
            - 127.0.0.1/8
            - 192.168.2./16

        # enables static file serving
        static: advanced
        # equals to:
        # ---
        # static:
        #     public_dir: '%kernel.project_dir%/public'
        #     strategy: advanced
        # ---
        # strategy can be one of: (default) auto, off, advanced, default
        #   - off: turn off feature
        #   - auto: use 'advanced' when debug enabled or not production environment
        #   - advanced: use request handler class \K911\Swoole\Server\RequestHandler\AdvancedStaticFilesServer
        #   - default: use default swoole static serving (faster than advanced, but supports less content types)

        # enables hot module reload using inotify
        hmr: auto
        # hmr can be one of: off, (default) auto, inotify
        #   - off: turn off feature
        #   - auto: use inotify if installed in the system
        #   - inotify: use inotify

        # enables api server on specific port
        # by default it is disabled (can be also enabled using --api flag via cli)
        api: true
        # equals to:
        # ---
        # api:
        #     enabled: true
        #     host: 0.0.0.0
        #     port: 9200

        # additional swoole symfony bundle services
        services:
            # see: \K911\Swoole\Bridge\Symfony\HttpKernel\DebugHttpKernelRequestHandler
            debug_handler: true

            # see: \K911\Swoole\Bridge\Symfony\HttpFoundation\TrustAllProxiesRequestHandler
            trust_all_proxies_handler: true

            # see: \K911\Swoole\Bridge\Symfony\HttpFoundation\CloudFrontRequestFactory
            cloudfront_proto_header_handler: true
            
            # this enables profiling using Blackfire (do not use in production)
            # see: \K911\Swoole\Bridge\Symfony\Profiling\BlackfireHandler
            # use HTTP HEADER x-blackfire-query with any value to enable profiling of the request
            # the header can be automatically added by using the blackfire extension for Chrome:
            # https://blackfire.io/docs/integrations/browsers/chrome
            blackfire_handler: true
    
            # this enables profiling of multiple requests after each other (might be useful for searching for memory leaks)
            # to activate profiling, a request with the GET parameter 'blackfire_start' needs to be made,
            # e.g. http://localhost/?blackfire_start=my_profile
            # after that, you can make multiple requests...
            # to stop profiling a new request has to be made, with the GET parameter 'blackfire_stop',
            # e.g. http://localhost/?blackfire_stop
            # after that, the blackfire profile will be uploaded to the blackfire servers, where you can inspect it
            # IMPORTANT!!! with this handler turned on, the swoole worker count and reactor count will get overridden to 1,
            # because the profiling needs to be ended in the same worker in which it has started
            blackfire_multi_request_handler: true
        
        # swoole http server settings
        # see https://www.swoole.co.uk/docs/modules/swoole-server/configuration
        settings:
            reactor_count: 2
            worker_count: 4
            # when not set, swoole sets these are automatically set based on count of host CPU cores
            task_worker_count: 2 # one of: positive number, "auto", or null to disable creation of task worker processes (default: null)

            log_level: auto
            # can be one of: (default) auto, debug, trace, info, notice, warning, error
            #   - auto: when debug set to debug, when not set to notice
            #   - {debug,trace,info,notice,warning,error}: see swoole configuration

            log_file: '%kernel.logs_dir%/swoole_%kernel.environment%.log'
            pid_file: /var/run/swoole_http_server.pid

            buffer_output_size: 2097152
            # in bytes, 2097152b = 2 MiB
  
            # default 0, if set then swoole worker gets restarted after after max_request + rand(0, max_request_grace)
            # requests
            max_request: 100
            # default max_request / 2, 
            max_request_grace: 50
    messenger_workers:
        # messenger workers need to be enabled explicitly
        enabled: true
```


src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php
<?php

declare(strict_types=1);

namespace K911\Swoole\Bridge\Symfony\Bundle\DependencyInjection;

use K911\Swoole\Common\Decoder;
use Symfony\Component\Config\Definition\Builder\TreeBuilder;
use Symfony\Component\Config\Definition\ConfigurationInterface;

final class Configuration implements ConfigurationInterface
{
    public const DEFAULT_PUBLIC_DIR = '%kernel.project_dir%/public';

    private const CONFIG_NAME = 'swoole';

    private $builder;

    public function __construct(TreeBuilder $builder)
    {
        $this->builder = $builder;
    }

    /**
     * {@inheritdoc}
     *
     * @throws \InvalidArgumentException
     * @throws \RuntimeException
     * @SuppressWarnings(PHPMD.ExcessiveMethodLength)
     */
    public function getConfigTreeBuilder(): TreeBuilder
    {
        $this->builder->getRootNode()
            ->children()
                ->enumNode('mode')
                    ->values(['transparent', 'strict'])
                    ->defaultValue('strict')
                ->end()
                ->arrayNode('http_server')
                    ->addDefaultsIfNotSet()
                    ->children()
                        ->scalarNode('host')
                            ->cannotBeEmpty()
                            ->defaultValue('0.0.0.0')
                        ->end()
                        ->scalarNode('port')
                            ->cannotBeEmpty()
                            ->defaultValue(9501)
                        ->end()
                        ->arrayNode('trusted_hosts')
                            ->defaultValue([])
                            ->prototype('scalar')->end()
                            ->beforeNormalization()
                                ->ifString()
                                ->then(function ($v): array { return Decoder::decodeStringAsSet($v); })
                            ->end()
                        ->end()
                        ->arrayNode('trusted_proxies')
                            ->defaultValue([])
                            ->prototype('scalar')->end()
                            ->beforeNormalization()
                                ->ifString()
                                ->then(function ($v): array { return Decoder::decodeStringAsSet($v); })
                            ->end()
                        ->end()
                        ->enumNode('running_mode')
                            ->cannotBeEmpty()
                            ->defaultValue('process')
                            ->values(['process', 'reactor', 'thread'])
                        ->end()
                        ->enumNode('socket_type')
                            ->cannotBeEmpty()
                            ->defaultValue('tcp')
                            ->values(['tcp', 'tcp_ipv6', 'udp', 'udp_ipv6', 'unix_dgram', 'unix_stream'])
                        ->end()
                        ->booleanNode('ssl_enabled')
                            ->defaultFalse()
                            ->treatNullLike(false)
                        ->end()
                        ->enumNode('hmr')
                            ->cannotBeEmpty()
                            ->defaultValue('auto')
                            ->treatFalseLike('off')
                            ->values(['off', 'auto', 'inotify'])
                        ->end()
                        ->arrayNode('api')
                            ->addDefaultsIfNotSet()
                            ->beforeNormalization()
                                ->ifTrue(function ($v): bool {
                                    return is_string($v) || is_bool($v) || is_numeric($v) || null === $v;
                                })
                                ->then(function ($v): array {
                                    return [
                                        'enabled' => (bool) $v,
                                        'host' => '0.0.0.0',
                                        'port' => 9200,
                                    ];
                                })
                            ->end()
                            ->children()
                                ->booleanNode('enabled')
                                    ->defaultFalse()
                                ->end()
                                ->scalarNode('host')
                                    ->cannotBeEmpty()
                                    ->defaultValue('0.0.0.0')
                                ->end()
                                ->scalarNode('port')
                                    ->cannotBeEmpty()
                                    ->defaultValue(9200)
                                ->end()
                            ->end()
                        ->end()
                        ->arrayNode('static')
                            ->addDefaultsIfNotSet()
                            ->beforeNormalization()
                                ->ifString()
                                ->then(function ($v): array {
                                    return [
                                        'strategy' => $v,
                                        'public_dir' => 'off' === $v ? null : self::DEFAULT_PUBLIC_DIR,
                                    ];
                                })
                            ->end()
                            ->children()
                                ->enumNode('strategy')
                                    ->defaultValue('auto')
                                    ->treatFalseLike('off')
                                    ->values(['off', 'default', 'advanced', 'auto'])
                                ->end()
                                ->scalarNode('public_dir')
                                    ->defaultValue(self::DEFAULT_PUBLIC_DIR)
                                ->end()
                            ->end()
                        ->end() // end static
                        ->arrayNode('exception_handler')
                            ->addDefaultsIfNotSet()
                            ->beforeNormalization()
                                ->ifString()
                                ->then(function ($v): array {
                                    return [
                                        'type' => $v,
                                        'verbosity' => 'auto',
                                        'handler_id' => null,
                                    ];
                                })
                            ->end()
                            ->children()
                                ->enumNode('type')
                                    ->cannotBeEmpty()
                                    ->defaultValue('production')
                                    ->treatFalseLike('production')
                                    ->values(['json', 'production', 'custom', 'auto'])
                                ->end()
                                ->enumNode('verbosity')
                                    ->cannotBeEmpty()
                                    ->defaultValue('auto')
                                    ->treatFalseLike('auto')
                                    ->values(['trace', 'verbose', 'default', 'auto'])
                                ->end()
                                ->scalarNode('handler_id')
                                    ->defaultNull()
                                ->end()
                            ->end()
                        ->end() // end exception_handler
                        ->arrayNode('services')
                            ->addDefaultsIfNotSet()
                            ->children()
                                ->booleanNode('debug_handler')
                                    ->defaultNull()
                                ->end()
                                ->booleanNode('trust_all_proxies_handler')
                                    ->defaultFalse()
                                    ->treatNullLike(false)
                                ->end()
                                ->booleanNode('cloudfront_proto_header_handler')
                                    ->defaultFalse()
                                    ->treatNullLike(false)
                                ->end()
                                ->booleanNode('entity_manager_handler')
                                    ->defaultNull()
                                ->end()
                                ->booleanNode('session_cookie_event_listener')
                                ->defaultFalse()
                                ->treatNullLike(false)
                                ->end()
                                ->booleanNode('blackfire_handler')
                                    ->defaultNull()
                                    ->treatNullLike(false)
                                ->end()
                                ->booleanNode('blackfire_multi_request_handler')
                                    ->defaultNull()
                                    ->treatNullLike(false)
                                ->end()
                            ->end()
                        ->end() // drivers
                        ->arrayNode('settings')
                            ->addDefaultsIfNotSet()
                            ->children()
                                ->scalarNode('log_file')
                                    ->cannotBeEmpty()
                                    ->defaultValue('%kernel.logs_dir%/swoole_%kernel.environment%.log')
                                ->end()
                                ->enumNode('log_level')
                                    ->cannotBeEmpty()
                                    ->values(['auto', 'debug', 'trace', 'info', 'notice', 'warning', 'error'])
                                    ->defaultValue('auto')
                                ->end()
                                ->scalarNode('pid_file')
                                    ->defaultNull()
                                ->end()
                                ->scalarNode('buffer_output_size')
                                    ->defaultValue(2097152)
                                ->end()
                                ->scalarNode('worker_count')
                                    ->defaultValue(1)
                                ->end()
                                ->scalarNode('reactor_count')
                                    ->defaultValue(1)
                                ->end()
                                ->scalarNode('task_worker_count')
                                    ->defaultNull()
                                ->end()
                                ->scalarNode('max_request')
                                    ->defaultNull()
                                ->end()
                                ->scalarNode('max_request_grace')
                                    ->defaultNull()
                                ->end()
                            ->end()
                        ->end() // settings
                    ->end()
                ->end() // server
                ->arrayNode('messenger_workers')
                    ->addDefaultsIfNotSet()
                    ->children()
                        ->booleanNode('enabled')
                            ->defaultFalse()
                            ->treatNullLike(false)
                        ->end()
                    ->end()
                ->end() // messenger workers
            ->end()
        ;

        return $this->builder;
    }

    public static function fromTreeBuilder(): self
    {
        $treeBuilderClass = TreeBuilder::class;

        return new self(new $treeBuilderClass(self::CONFIG_NAME));
    }
}
